apache2/msc_reqbody.c

```diff
@@ -428,55 +428,62 @@ apr_status_t modsecurity_request_body_store(modsec_rec *msr,
-        first_pkt = 1;
-    }
-    else    {
-
-        data = (char *)malloc(msr->stream_input_length + 1 - buflen);
+    if (msr->stream_input_data == NULL)  {
+        // Is the request body length is known beforehand? (requests that are not Transfer-Encoding: chunked)
+        if (msr->request_content_length > 0) {
+            allocate = msr->request_content_length;
+        }
+        else {
+            // We don't know how this request is going to be, so hope for just buflen to begin with (requests that are Transfer-Encoding: chunked)
+            allocate = buflen;
+        }
 
-        if(data == NULL)
+        allocated = (char*) calloc(allocate, sizeof(char));
+        if (allocated) {
+            msr->stream_input_data = allocated;
+            msr->stream_input_allocated_length = allocate;
+        }
+        else {
+            *error_msg = apr_psprintf(
+                msr->mp,
+                "Unable to allocate memory to hold request body on stream. Asked for %" APR_SIZE_T_FMT " bytes.",
+                allocate);
             return -1;
-
-        memset(data, 0, msr->stream_input_length + 1 - buflen);
-        memcpy(data, msr->stream_input_data, msr->stream_input_length - buflen);
-
-        stream_input_body = (char *)realloc(msr->stream_input_data, msr->stream_input_length + 1);
-
-        msr->stream_input_data = (char *)stream_input_body;
-    }
-
-    if (msr->stream_input_data == NULL) {
-        if(data)    {
-            free(data);
-            data = NULL;
         }
-        *error_msg = apr_psprintf(msr->mp, "Unable to allocate memory to hold request body on stream. Asked for %" APR_SIZE_T_FMT " bytes.",
-                msr->stream_input_length + 1);
-        return -1;
     }
+    else {
+        // Do we need to expand the space we have previously allocated?
+        if ((msr->stream_input_length + buflen) > msr->stream_input_allocated_length) {
 
-    memset(msr->stream_input_data, 0, msr->stream_input_length+1);
+            // If this becomes a hotspot again, consider increasing by some percent extra each time, for fewer reallocs
+            allocate = msr->stream_input_length + buflen;
 
-    if(first_pkt)   {
-        memcpy(msr->stream_input_data, buffer, msr->stream_input_length);
-    } else {
-        memcpy(msr->stream_input_data, data, msr->stream_input_length - buflen);
-        memcpy(msr->stream_input_data+(msr->stream_input_length - buflen), buffer, buflen);
+            allocated = (char*) realloc(msr->stream_input_data, allocate);
+            if (allocated) {
+                msr->stream_input_data = allocated;
+                msr->stream_input_allocated_length = allocate;
+            }
+            else {
+                *error_msg = apr_psprintf(
+                    msr->mp,
+                    "Unable to reallocate memory to hold request body on stream. Asked for %" APR_SIZE_T_FMT " bytes.",
+                    allocate);
+                free(msr->stream_input_data);
```

Comment:
it won't make much difference. A crash is imminent at this point. We can do it for consistency though, in case a way for recovery becomes possible in the future.